conbench/entities/benchmark_result.py

```diff
@@ -138,168 +142,116 @@ def create(userres):
-        # iteration samples. Rely on `stats` key to exist (see checks above).
-        elif "stats" in userres and do_iteration_samples_look_like_error(
-            userres["stats"]["data"]
-        ):
-            # User did not set `error`, but the number sequence indicates
-            # that this result is to be considered 'errored'. Set a generic
-            # error message.
-            result_error_for_db = {
+            # We have business logic elsewhere that checks only for presence of
+            # the `error` key (ignores its value, a value of `None` might
+            # elsewhere be interpreted as error -- this did cost me 30 minutes
+            # of debugging).
+            result_data_for_db["error"] = userres["error"]
+
+        # Check for a more subtle error condition based on the per-iteration
+        # samples. Invariant: if "error" is not present then "stats" is present
+        # as a key in this dictionary -- this is schema-enforced. he `stats`
+        # object is guaranteed to have a `data` key.
+        elif do_iteration_samples_look_like_error(userres["stats"]["data"]):
+            # User indicated error with variant B: missing or incomplete data.
+            # User unfortunately did not set `error` explicitly, but we err on
+            # the side auf caution here and treat the result as 'errored'. This
+            # is documented. Set generic error detail.
+            result_data_for_db["error"] = {
+                # Maybe tune this error message to be more generic.
                 "status": "Partial result: not all iterations completed"
             }
 
-        # This is going to be the structure based on which we will do
-        # DB insertion.
-        benchmark_result_data = {}
-        # Only insert `error` key if there is an error object. In JSON, error:
-        # null still means that the benchmark is errored (at least some
-        # business logic treat it as such.)
-        if result_error_for_db is not None:
-            benchmark_result_data["error"] = result_error_for_db
-
-        # Temporary: keep name `data` -- it's unfortunate that we have the name
-        # `data` here while also having key(s) in the dict(s) that are called
-        # `data`. Assign a recognizable name to the big, outest, user-given
-        # object: userres is not ideal, but a start.
-        data = userres
+        else:
+            # process_samples_build_agg() must only be called if
+            # do_iteration_samples_look_like_error() returned False. That's
+            # the case here.
+            result_data_from_stats = validate_and_aggregate_samples(userres["stats"])
+
+            # Per-iteration samples looked good, and we did (potentially)
+            # rebuild aggregates. Merge dict `result_stats_data_for_db` on top
+            # of dict `benchmark_result_data`, overwriting upon conflict.
+            result_data_for_db |= result_data_from_stats  # PEP 584 update
```

Comment:
change in behavior around here: the aggregates (mean, stdev, etc...) now _take precedence_ over user-given data.

See below: a warning is logged for mismatch.

For data analysis in Conbench I'd absolutely like to work with the guarantee that if there are N data points and a stdev value that the stdev value is _consistent_ with the data points. Trusting user-given data in that regard doesn't provide that guarantee.

This is very much related to / about https://github.com/conbench/conbench/issues/533

Reply:
When I implemented server-side stats calculations, I believe @jonkeane was against them taking precedence over user-supplied values, but I'm trying to think of a case where they're not what we want and coming up blank. Most cases I can think of have switched to not sending them anyway.

Reply:
This demands careful consideration, indeed. I think this is reasonably easy to see through. We are in a code section with two invariants:

1) at least three data points ("samples")
2) each one is a numeric value

From here, we have three options:

1) do not validate consistency between samples and aggregates
2) validate, and upon error reject reject
3) validate, and upon error "take the more correct outcome"

I think (3) is more or less obviously the best option. It is pretty much the same as ignoring the user-provided aggregates. But again, this given the two invariants above.

Yes, a user should be to inject arbitrary aggregates, but not when _also_ providing "many" individual samples. In my view, these two should be mutually exclusive. 

We need to draw a clear line between:

a) here are N samples, _you_ do the aggregates
b) here are the aggregates, and I specifically _don't_ give you the underlying raw samples -- 

I think drawing this line can be easy to do, especially when starting from scratch. Given what we have, it takes a bit of time, changing API behavior, and changing API docs to massage things into the direction where we have these two clearly different interface use cases. Here, we work on (a) and make it well-specified. (b) still needs a bit of work, mainly specification work.

>  @jonkeane was against them taking precedence over user-supplied values

I hope this was about implementing (b).

> Most cases I can think of have switched to not sending them anyway.

Thanks for confiming!

Reply:
Thank you for the feedback @austin3dickey!